sympy/solvers/ode/ode.py

```diff
@@ -731,6 +731,222 @@ def _helper_simplify(eq, hint, match, simplify=True, ics=None, **kwargs):
+
+    >>> from sympy import symbols, dsubs, Function, cos
+    >>> f, g, h = symbols('f g h', cls=Function)
+    >>> x, a, t = symbols('x a t')
+    >>> eq = f(x).diff(x) + f(x)
+    >>> dsubs(eq, {x: t**2})
+    f(t**2) + Derivative(f(t**2), t)/(2*t)
+
+    Functions in the equation can be transformed too
+
+    >>> eq = f(x).diff(x, 2) + cos(x)
+    >>> dsubs(eq, {x: t**2, f(x): g(t)})
+    cos(t**2) + (Derivative(g(t), (t, 2))/(2*t)
+    - Derivative(g(t), t)/(2*t**2))/(2*t)
+    >>> dsubs(eq, {x: t, f(x): g(t)**2})
+    2*g(t)*Derivative(g(t), (t, 2)) + cos(t)
+    + 2*Derivative(g(t), t)**2
+
+    To apply a transformation with number of new variables greater
+    than number of old variables, a list of all new variables must be
+    given. For example, a transformation `x = a t` would be done as -
+
+    >>> dsubs(eq, {x: a*t, f(x): g(t)}, [t])
+    cos(a*t) + Derivative(g(t), (t, 2))/a**2
+    >>> dsubs(eq, {x: t, f(x): h(a)*g(t)**2}, [t, g(t)])
+    2*g(t)*h(a)*Derivative(g(t), (t, 2))
+    + 2*h(a)*Derivative(g(t), t)**2 + cos(t)
+
+    Let us see how ``dsubs`` works for some typical cases using an example
+    of homogeneous ODEs which can be solved using the substitution
+    `v = \frac{y}{x}`.
+
+    >>> from sympy import symbols, Function, dsolve
+    >>> x, t = symbols('x t')
+    >>> y, v = symbols('y v', cls=Function)
+    >>> eq = y(x).diff(x) - (x**2 + y(x)**2)/(x*y(x))
+
+    Transform the homogeneous equation to a linear equation
+
+    >>> homeq = dsubs(eq, {x:t, y(x): t*v(t)}).simplify()
+    >>> homeq
+    t*Derivative(v(t), t) - 1/v(t)
+
+    Find the solutions for the transformed equation
+
+    >>> solt = dsolve(homeq)
+    >>> solt
+    [Eq(v(t), -sqrt(C1 + 2*log(t))), Eq(v(t), sqrt(C1 + 2*log(t)))]
+
+    Re-substitute `v = y x` to get the solutions to the
+    original equation
+
+    >>> sol = list(map(lambda sol: dsubs(sol, {v(t): y(x)/x, t: x}, reverse=True), solt))
+    >>> sol
+    [Eq(y(x)/x, -sqrt(C1 + 2*log(x))), Eq(y(x)/x, sqrt(C1 + 2*log(x)))]
+
+    Let us look at another example - The 2nd order Cauchy Euler equation.
+
+    >>> x, a, b, t = symbols('x a b t')
+    >>> y, phi = symbols('y phi', cls=Function)
+    >>> eq = x**2*y(x).diff(x, 2) + a*x*y(x).diff(x) + b*y(x)
+
+    Transform the equation using the substitutions `x = e^t`
+    and `y(x) = \phi(t)`
+
+    >>> from sympy import exp, log, logcombine
+    >>> eqtrans = dsubs(eq, {x: exp(t), y(x): phi(t)}).simplify()
+    >>> eqtrans
+    a*Derivative(phi(t), t) + b*phi(t) - Derivative(phi(t), t) + Derivative(phi(t), (t, 2))
+
+    The equation is now transformed to a constant-coefficient equation.
+
+    >>> solt = dsolve(eqtrans)
+    >>> solt
+    Eq(phi(t), C1*exp(t*(-a - sqrt(a**2 - 2*a - 4*b + 1) + 1)/2) + C2*exp(t*(-a + sqrt(a**2 - 2*a - 4*b + 1) + 1)/2))
+
+    Re-substitute for `\phi(t)` and `t` to get the solution for
+    the original equation.
+
+    >>> sol = logcombine(dsubs(solt, {phi(t): y(x), t: log(x)}, reverse=True), force=True)
```

Comment:
The idea with `rerverse` was supposed to be that you would give it the same arguments as the forward substitution. If you have to pass in a different transformation dict then is there any point in having reverse?

Reply:
So, we use `solve` inside?

Reply:
We don't want to use `solve` because it tries to find all solutions in cases where there isn't a unique solution. Here we just want a local inverse. Rather than use `solve` I would prefer to require the user to supply the reverse transformation perhaps like:
```python
trans = ({t: exp(tau), x(t):x(tau)}, {tau:log(t), x(tau):x(t)})

eqtrans = dsubs(eq, trans)
soltrans = dsolve(eqtrans)
sol = dsubs(soltrans, trans, reverse=True)
```
Maybe some entries could be automatically filled like `x(tau):x(t)` if not given in the reverse part of the transformation. This approach would also mean that `newvars` is not needed because we can see which are the new variables by looking at the keys of the reverse transformation dict.

If we aren't using the same arguments though then there isn't any point in having `reverse=True` because the user should be able to use `dsubs` in both directions:
```python
trans = {t: exp(tau), x(t):x(tau)}
trans_reverse = {tau:log(t), x(tau):x(t)}

eqtrans = dsubs(eq, trans)
soltrans = dsolve(eqtrans)
sol = dsubs(soltrans, trans_reverse)
```
Maybe it would just be simpler to do that.

Reply:
> This approach would also mean that `newvars` is not needed because we can see which are the new variables by looking at the keys of the reverse transformation dict.

I think `newvars` is still required for cases like `{x: a*t, f(x): h(a)*g(t)}`.

> If we aren't using the same arguments though then there isn't any point in having reverse=True because the user should be able to use dsubs in both directions:

So, I remove the `reverse=True`? The present code actually works without it, although I'd have to test a few more examples to make sure everything is happening correctly

Reply:
> I think `newvars` is still required for cases like `{x: a*t, f(x): h(a)*g(t)}`.

You could specify that as
```python
trans_forward  = {x: a*t, f(x): h(a)*g(t)}
trans_backward = {t: x/a, g(t): f(x)/h(a)}
transformation = (trans_forward, trans_backward)

eq_subs = dsubs(eq, transformation)
```
Then `dsubs` knows that `t` and `g(t)` are the new variables because they are the keys in the backward transformation dict. I think that means that `newvars` is not needed.

It's also possible that the transformation could be computed differently if the backward transformation is also supplied. The docs for Maple's `dchange` imply that Maple uses a different algorithm in that case.

Reply:
> Then `dsubs` knows that `t` and `g(t)` are the new variables because they are the keys in the backward transformation dict. I think that means that `newvars` is not needed.

Okay, I get it now. I will change this to be the interface.

> It's also possible that the transformation could be computed differently if the backward transformation is also supplied. 

How can it be computed differently? Could you elaborate?

Reply:
> > I think `newvars` is still required for cases like `{x: a*t, f(x): h(a)*g(t)}`.
> 
> You could specify that as
> 
> ```python
> trans_forward  = {x: a*t, f(x): h(a)*g(t)}
> trans_backward = {t: x/a, g(t): f(x)/h(a)}
> transformation = (trans_forward, trans_backward)
> 
> eq_subs = dsubs(eq, transformation)
> ```
> 
> Then `dsubs` knows that `t` and `g(t)` are the new variables because they are the keys in the backward transformation dict. I think that means that `newvars` is not needed.
> 
> It's also possible that the transformation could be computed differently if the backward transformation is also supplied. The docs for Maple's `dchange` imply that Maple uses a different algorithm in that case.

About this interface, I was thinking, the user needs to specify both the directions twice. It doesn't seem nice to me. Is that okay with you?

Reply:
> About this interface, I was thinking, the user needs to specify both the directions twice. It doesn't seem nice to me. Is that okay with you?

I don't know what you mean by "both the directions twice".

The point is that we can make this work with only the forward transformation but it leaves us needing extra information such as `newvars` and that extra information is actually much less useful than the reverse transformation that can be used to compute the result with a different algorithm. If we have the user supply the reverse transformation then we can implement better algorithms later.

Not providing the reverse transformation might seem nice in interactive usage but in actual practice when would you be able to do a meaningful calculation without also needing to have the reverse transformation as well?

If the caller already needs to have the reverse transformation then why shouldn't they give it to `dsubs` since it makes the signature of `dsubs` unambiguous? For a caller who already has the transformation in both directions asking them to supply `newvars` instead is just annoying.

If the interface is not that friendly then we can make a new function later with a nicer UI (that just calls `dsubs` internally). I think it would also be reasonable to design this now so that it could later be changed to make the reverse transformation optional. We should start from designing something that is clearly well defined and useful for library use before considering how to make it more convenient for e.g. interactive use.

Reply:
> I don't know what you mean by "both the directions twice".

By this I meant we are passing in both the original and reverse transformations into `dsubs` once during the forward substitution and once during the reverse substitution. But since you're saying we can modify the interface later, I will change the code and make the interface as you suggested above.

Reply:
I've made the requested changes